crates/sui-core/src/streamer.rs

```diff
@@ -1,21 +1,96 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use sui_types::event::EventEnvelope;
-use tokio::{sync::mpsc::Receiver, task::JoinHandle};
-use tracing::debug;
 
-pub struct Streamer {
-    event_queue: Receiver<EventEnvelope>,
+use crate::event_handler::EVENT_DISPATCH_BUFFER_SIZE;
+use futures::Stream;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use sui_types::base_types::ObjectID;
+use sui_types::error::SuiError;
+use tokio::runtime::Handle;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{mpsc, RwLock};
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::{debug, warn};
+
+type Subscribers<T> = BTreeMap<String, Sender<T>>;
+type FilterMap<F, T> = Arc<RwLock<BTreeMap<F, T>>>;
```

Comment:
Do we need a RwLock here?   I was just thinking that once you subscribe you should not be able to change the filter.  Just for simplicity.   Or is the RwLock fort adding more filters to the map?

We could think about simplifying this by just having a list of subscribers each with a filter:

    struct SubscriberInfo<T> {
      sender: Sender<T>,
      filters: EventFilter,
    }
    type Subscribers<T> = Vec<(String, SubscriberInfo)>;